pyfr/solvers/base/elements.py

```diff
@@ -211,47 +211,87 @@ def _gen_pnorm_fpts(self):
+
+        return ploc
+
+    @lazyprop
+    def _smats_mpts(self):
+        # Metric basis with grid point (q<=p) or pseudo grid points (q>p)
+        mpts = self.basis.mpts
+        mbasis = self.basis.mbasis
+        self.nmpts = nmpts = len(mpts)
+
+        # Dimensions and number of elements
+        ndims = self.ndims
+        neles = self.neles
+
+        # Coordinate at pts
+        x = self._ploc_at_pts_np(mpts)
+
+        # Jacobian at pts
+        jacop = np.rollaxis(mbasis.jac_nodal_basis_at(mpts), 2)
+        jacop = jacop.reshape(-1, nmpts)
 
         # Cast as a matrix multiply and apply to eles
-        jac = np.dot(jacop.reshape(-1, nspts), self.eles.reshape(nspts, -1))
+        jac = np.dot(jacop, x.reshape(nmpts, -1))
 
         # Reshape (npts*ndims, neles*ndims) => (npts, ndims, neles, ndims)
-        jac = jac.reshape(npts, ndims, neles, ndims)
+        jac = jac.reshape(nmpts, ndims, ndims, neles)
 
         # Transpose to get (ndims, npts, ndims, neles)
-        return jac.transpose(3, 0, 1, 2)
+        jac = jac.transpose(2, 0, 1, 3)
 
-    def _get_smats(self, pts, retdets=False):
-        jac = self._get_jac_eles_at(pts)
         smats = np.empty_like(jac)
 
-        if self.ndims == 2:
+        if ndims == 2:
+            # Just compute smats whose order is q-1
             a, b, c, d = jac[0,:,0], jac[0,:,1], jac[1,:,0], jac[1,:,1]
 
             smats[0,:,0], smats[0,:,1] = d, -b
             smats[1,:,0], smats[1,:,1] = -c, a
 
-            if retdets:
-                djacs = a*d - b*c
+            self._djacs_mpts = a*d - b*c
         else:
-            # We note that J = [x0, x1, x2]
-            x0, x1, x2 = jac[:,:,0], jac[:,:,1], jac[:,:,2]
+            # Cpmpute x cross x_(chi)
+            jac = np.rollaxis(jac, 2)
+            tt = [np.cross(x, dx, axisa=1, axisb=0, axisc=1) for dx in jac]
+            tt = np.array(tt).reshape(ndims, nmpts, -1)
 
-            smats[0] = np.cross(x1, x2, axisa=0, axisb=0, axisc=1)
-            smats[1] = np.cross(x2, x0, axisa=0, axisb=0, axisc=1)
-            smats[2] = np.cross(x0, x1, axisa=0, axisb=0, axisc=1)
+            # Derivative of x cross x_(chi) at (pseudo) grid points
+            dtt = np.array([np.dot(jacop, tn) for tn in tt])
+            dtt = dtt.reshape(ndims, nmpts, ndims, ndims, -1).swapaxes(1, 2)
 
-            if retdets:
-                # Exploit the fact that det(J) = x0 . (x1 ^ x2)
-                djacs = np.einsum('ij...,ji...->j...', x0, smats[0])
+            # Kopriva's invariant form of smats
+            # Ref. JSC 26(3), 301-327, Eq. (37)
+            smats[0] = 0.5*(dtt[2][1] - dtt[1][2])
+            smats[1] = 0.5*(dtt[0][2] - dtt[2][0])
+            smats[2] = 0.5*(dtt[1][0] - dtt[0][1])
 
-        return (smats, djacs) if retdets else smats
+            # Exploit the fact that det(J) = x0 . (x1 ^ x2)
+            self._djacs_mpts = np.einsum('ij...,ji...->j...', jac[0], smats[0])
```

Comment:
Adding some extra lines is fine.  The key is to remove the side effect (which is an assignment inside of the function).